Changes for edge-21.1.3 (#5590)
linkerd/linkerd2: CHANGES.md

# Changes

## edge-21.1.3

> This edge release improves proxy diagnostics and handling for the failfast state

Reviewer: Probably most readers are unfamiliar with what failfast means.  What about something like: "... improves proxy diagnostics and recovery in situations where the proxy is temporarily unable to route requests". ? @olix0r wdyt?

Reviewer 2: +1

by improving log output and enhancing the proxy's ability to recover when it is

> in failfast. In addition, the commands and sub-commands for `linkerd viz` have
> been updated to be consistent with the `linkerd` core commands. Also, the
> install commands for the core and multicluster components have been modified to
> accept helm-style customization flags.

Reviewer 2: This is a little unclear to me. Perhaps a higher-level summary like:

```suggestion
in failfast. Additionally, the `viz` and `multicluster` CLI sub-commands have
been updated for consistency. 
```

Full release notes:

> * Added helm `set`, `set-string`, `values`, `set-files` customization flags for

Reviewer: s/helm/Helm-style

  the `linkerd install` and `linkerd multicluster install` commands

> * Fixed `linkerd metrics` resource selector to differentiate between owner

Reviewer: Fixed an issue where `linkerd metrics` could return metrics for the incorrect set of pods when there are overlapping label selectors

  resource names
* Added tap-injector to linkerd-viz which is responsible for adding the tap
  service name environment variable to the Linkerd proxy container

> * Improved proxy diagnostics about the proxy's failfast state

Reviewer 2: ```suggestion
* Improved diagnostics about the proxy's failfast state
```

* Made failfast recovery for a service more robust when new requests are being
  received

> * Improved metric labeling for TCP server metrics, as well as inbound and

Reviewer: Improved how?

Reviewer 2: We can probably omit this. It's not user-facing in any of our tools yet.

>   outbound tls metrics

Reviewer 2: ```suggestion
  outbound TLS metrics
```

> * Added `client` and `server` prefixes to socket-level errors to indicate which

Reviewer: I'm not sure what this refers to.  Log messages? Metric labels?

  side of the proxy encountered the error

> * Added a watch to jaeger-injector to prevent errors in environments with high

Reviewer: Added watch RBAC permissions to the jaeger-injector to prevent...

> * Added a watch to jaeger-injector to prevent errors in environments with high
>   pod or namespace churn

Reviewer 2: Perhaps
```suggestion
* Improved jaeger-injector reliability in environments with many resources
```

> * Added check to confirm whether the jaeger injector pod is in running state

Reviewer: jaeger-injector

  (thanks @yashvardhan-kukreja!)

> * Fixed some logic that caused a crash when EndpointSlices were enabled
>   (thanks @oleh-ozimok!)

Reviewer 2: ```suggestion
* Fixed a crash in the destination controller when EndpointSlices were enabled
  (thanks @oleh-ozimok!)
```

* Added a `linkerd viz check` sub-command to verify the states of the
  `linkerd-viz` components
* Added a `log-format` flag to optionally output the control plane component log
  output as JSON (thanks @mo4islona!)
* Added logic to use the configured default namespace for `metrics` and

>   `profile` subcommands

Reviewer 2: Not clear to me. Did we fix a problem in the prior edge? If so, perhaps "Fixed an issue where the `metrics` and `profile` subcommands used the incorrect viz namespace"?
